flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -540,15 +540,12 @@ private boolean performCheckpoint(CheckpointMetaData checkpointMetaData) throws
 
 		synchronized (lock) {
 			if (isRunning) {
+				checkpointState(checkpointMetaData);
 
-				// Since both state checkpointing and downstream barrier emission occurs in this
-				// lock scope, they are an atomic operation regardless of the order in which they occur.
-				// Given this, we immediately emit the checkpoint barriers, so the downstream operators
-				// can start their checkpoint work as soon as possible
+				// broadcast barriers after snapshot operators' states.
 				operatorChain.broadcastCheckpointBarrier(
-						checkpointMetaData.getCheckpointId(), checkpointMetaData.getTimestamp());
-
-				checkpointState(checkpointMetaData);
+						checkpointMetaData.getCheckpointId(), checkpointMetaData.getTimestamp()
+				);
```

Comment:
Disabling chain is an option, but it is a little strict to force it. Users may try to avoid network overhead by chaining them together.
I think it is worth to have a try to introduce `ReentrantReadWriteLock`. Maybe I can have a test between `Object.lock()` and `ReentrantReadWriteLock`, to get the performance difference while locking and unlock for many times.